library/modules/EventManager.cpp

```diff
@@ -479,17 +596,82 @@ TODO: consider checking item creation / experience gain just in case
 static void manageJobCompletedEvent(color_ostream& out) {
     if (!df::global::world)
         return;
-    static int32_t last_tick = -1;
+    static int32_t fn_last_tick = -1;
     int32_t tick = df::global::world->frame_counter;
+    int32_t oldest_last_tick = (uint16_t)-1;
 
-    multimap<Plugin*,EventHandler> copy(handlers[EventType::JOB_COMPLETED].begin(), handlers[EventType::JOB_COMPLETED].end());
-    map<int32_t, df::job*> nowJobs;
-    for ( df::job_list_link* link = df::global::world->jobs.list.next; link != NULL; link = link->next ) {
-        if ( link->item == NULL )
-            continue;
-        nowJobs[link->item->id] = link->item;
+    // update the completed jobs list (this apparently isn't as straight forward as one might think)
+    // todo: verify the need for this check
+    //if it happened within a tick, must have been cancelled by the user or a plugin: not completed
+    if(tick > fn_last_tick) {
+        // we're going to need to check if started jobs are still on this list, so we copy the IDs to avoid a linear search into this list for each started_job
+        std::unordered_map<int32_t, df::job*> current_jobs;
+        for (df::job_list_link* link = df::global::world->jobs.list.next; link != NULL; link = link->next) {
+            df::job* job = link->item;
+            current_jobs.emplace(job->id, job);
+        }
+        // iterate the started jobs list
+        for (auto &iter: startedJobs) {
```

Comment:
Seems as though `startedJobs` is insufficient for this task. Testing shows this manager doesn't work correctly and this is the only thing that makes a lot of sense.